### PR TITLE
Call for creating an image from a disk

### PR DESCRIPTION
**Pull request by cbastuck**
Probably the function fits better in a different scope? On the other hand the path '/global/images' matches in project scope intrinsically.

Fixes #<issue_number_goes_here> (it's a good idea to open an issue first for discussion)
It's probably quicker to start a discussion from this PR?

- [(*)] Tests and linter pass
no new issues introduced, already on master: 6 problems (6 errors, 0 warnings)
- [ ] Code coverage does not decrease (if any source code was changed)
no tests have yet been added
- [*] Appropriate docs were updated (if necessary)

### CODE REVIEW

**Comment by cbastuck** [comment]
CLA was signed

**Comment by stephenplusplus** [comment]
Thanks for the PR. I think it should definitely be scoped where you put it (on `Project`), although I could see it being helpful on `Disk` as well. What do you think?

File: src/project.js [hunk 1]
```diff
@@ -107,6 +108,61 @@ function Project(compute) {
...
+    options
+  );
+
+  this.request(
+    {
+      method: 'POST',
+      uri: '/global/images',
+      json: body,
+    },
+    function(err, resp) {
```
Comment (stephenplusplus) [inline src/project.js]: Just passing in `callback` here will work.

File: src/project.js [hunk 2]
```diff
@@ -107,6 +108,61 @@ function Project(compute) {
...
+ * }
+ *
+ * project.createImage('image1', disk, callback);
+ *
+ * //-
+ * // If the callback is omitted, we'll return a Promise.
+ * //-
+ */
+Project.prototype.createImage = function(imageName, disk, options, callback) {
+  var body = extend(
```
Comment (stephenplusplus) [inline src/project.js]: We should add a throw if `disk` is not a Disk:

```js
if (!util.isCustomType(disk, 'Disk')) {
  throw new Error('A Disk object is required.');
}
```

File: src/project.js [hunk 3]
```diff
@@ -107,6 +108,61 @@ function Project(compute) {
 
 util.inherits(Project, common.ServiceObject);
 
+/**
+ * Create a image from disk in project.
+ *
```
Comment (stephenplusplus) [inline src/project.js]: Let's add a resource link, please:

```js
* @see [Images: insert API Documentation]{@link https://cloud.google.com/compute/docs/reference/v1/images/insert}
```

File: src/project.js [hunk 4]
```diff
@@ -107,6 +108,61 @@ function Project(compute) {
...
+ * @param {string} imageName - Name of the target image.
+ * @param {Disk} disk - The source disk to create the image from.
+ * @param {object} options - Further options, for details please refer to
```
Comment (stephenplusplus) [inline src/project.js]: To mark this as optional: `[options]`

File: src/project.js (hunk 1 above)
Comment (cbastuck) [reply src/project.js]: done

File: src/project.js (hunk 2 above)
Comment (cbastuck) [reply src/project.js]: done

File: src/project.js (hunk 3 above)
Comment (cbastuck) [reply src/project.js]: done

File: src/project.js (hunk 4 above)
Comment (cbastuck) [reply src/project.js]: done

**Comment by cbastuck** [comment]
Thanks for the quick reply and your suggestions.  

Having the call as well on Disk as well sounds good. On the other hand the disk relates to a Zone, which contradicts the 'global' concept. Thus, maybe better leave it as is, to avoid ambiguity?

**Comment by stephenplusplus** [comment]
> Thus, maybe better leave it as is, to avoid ambiguity?

👍 SGTM

**Comment by cbastuck** [comment]
Looks like node4 is a bit flaky on CI for 'Compute' test case? 
I did not investigate further but noticed the node failed before this patch (same timeout reported)

**Comment by stephenplusplus** [comment]
Yeah, don't worry about the CI. I have a commit coming soon. Before I send it, we have to figure out this CLA thing. Can you try commenting "I signed it!" to see if that'll wake up the CLA bot?

**Comment by cbastuck** [comment]
I signed it!

**Comment by cbastuck** [comment]
thanks for the hint, this re-invoked the bot after adding my github name (which was optional and therefore left blank)

**Comment by stephenplusplus** [comment]
So, I made another change and moved the new method from `Project` to `Compute`. (Sorry, it's been a while since I've worked with this API, so I've been a little slow to recall our conventions).

Most of our methods create or return a custom class, such as Disk. We generally expose `create{Object}`, `get{Object}s`, and then the custom class has methods to modify it: `{Object}.delete()`, `{Object}.setMetadata()`, etc.

What we should *really* do is create the `Image` class that has methods that match the [upstream API](https://cloud.google.com/compute/docs/reference/rest/v1/images), like `image.deprecate()`, `image.getMetadata()`, `image.setLabels()`, etc.

I'm okay with not implementing those for the time being. However, before I merge, would you be interested in creating them? `image.js` file would basically be a clone of any of our other files, such as `project.js`. Then the accessor methods, like `compute.getImages()`, and `compute.image()` would be clones of `compute.getFirewalls()` and `compute.firewall()` for example.

Let me know what you think!

**Comment by cbastuck** [comment]
I signed it!

**Comment by cbastuck** [comment]
Creating the `Image` class makes totally sense. I will look into it

**Comment by stephenplusplus** [comment]
Hey @cbastuck, thanks for doing all of this! Sorry I couldn't review sooner. I think it looks good. There might be a few trivial, convention things I'll take care of. I'll try to do this within the next day or so, and we can release right after.

**Comment by cbastuck** [comment]
Hi @stephenplusplus sounds good. Do I have to do something regarding the CLA? Maybe re-authoring all commits with the email address I signed the CLA with? When signing I added my github username which links to the 'authored' address: mail@cbastuck.de. Probably this is not resolved?

**Comment by stephenplusplus** [comment]
Actually, I think everything is fine with the bot. It has a known bug where it gets confused when a PR has two authors. Since you were fine before I pushed my commit, I believe we're good. (@alexander-fenster can you confirm?)

**Comment by alexander-fenster** [comment]
Could you please add tests to improve the coverage?

**Comment by alexander-fenster** [comment]
As for the bot, everything is OK as everyone have signed the CLA and as long as all parties are OK with the changes from all parties we are good to go.

> We need to confirm that all authors are ok with their commits being contributed to this project. Please have them confirm that here in the pull request.

So if all authors sign off the PR we can merge.

**Comment by cbastuck** [comment]
I will add tests

**Comment by cbastuck** [comment]
I was completely blocked be the daily work and did not find the time to increase the test coverage as promised, sorry.

**Comment by stephenplusplus** [comment]
No problem!

**Pull request merged**